File: 0130-surrounded-regions/0130-surrounded-regions.cpp
class Solution {
    private:
    void dfs(int i, int j,vector<vector<int>> &vis,vector<vector<char>>& board){
        int n=board.size();
        int m=board[0].size();
        vis[i][j]=1;
        int a[]={-1,0,1,0};
        int b[]={0, -1, 0, 1};
        for(int k=0; k<4 ;k++){
            int nr= i + a[k];
            int nc = j+ b[k];
            if(nr<0 || nc <0 || nr==n || nc==m ) continue;
            if(board[nr][nc]=='O' && !vis[nr][nc]) dfs(nr,nc,vis,board);
        }
    }
public:
    void solve(vector<vector<char>>& board) {
        int n=board.size();
        int m=board[0].size();

        vector<vector<int>> vis(n, vector<int>(m,0));

        for(int j=0;j<m;j++){
            if(board[0][j]=='O' && !vis[0][j]) dfs(0,j,vis,board);
            if(board[n-1][j]=='O' && !vis[n-1][j]) dfs(n-1, j, vis, board);
        }
        for(int i=0;i<n;i++){
            if(board[i][0]=='O' && !vis[i][0]) dfs(i,0,vis,board);
            if(board[i][m-1]=='O' && !vis[i][m-1]) dfs(i,m-1,vis,board);
        }

        for(int i=0; i<n ;i++){
            for(int j=0;  j<m ;j++){
                if(board[i][j]=='O' && !vis[i][j]) board[i][j]='X';
            }
        }
    }
};

File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    public void solve(char[][] board) {
      int r=board.length;
      int c=board[0].length;
      for(int i=0;i<r;i++){
        dfs(i,0,board,r,c);
        dfs(i,c-1,board,r,c);
      }
      for(int i=0;i<c;i++){
        dfs(0,i,board,r,c);
        dfs(r-1,i,board,r,c);
      }
      for(int i=0;i<r;i++){
        for(int j=0;j<c;j++){
            if(board[i][j]=='O'){
                board[i][j]='X';
            }
            if(board[i][j]=='A'){
                board[i][j]='O';
            }
        }
      }

    }
    public void dfs(int r,int c,char[][] board,int R,int C) {
        if(r<0 || r>=R || c<0 || c>=C || board[r][c]=='X' ||board[r][c]=='A' )return;
        board[r][c]='A';
        dfs(r+1,c,board,R,C);
        dfs(r-1,c,board,R,C);
        dfs(r,c+1,board,R,C);
        dfs(r,c-1,board,R,C);
    }
}